File: app/javascript/controllers/dns_check_controller.js
import { Controller } from "@hotwired/stimulus"

// Connects to data-controller="dns-check"
export default class extends Controller {
  static targets = ["domain", "table"]

  async check(event) {
    event.preventDefault()
    const domain = this.domainTarget.value
    const response = await fetch(`/dns_check/check?domain=${domain}`)
    const data = await response.json()

    console.log("CNAME records:", data.cname_records);

    const aRows = data.a_records.map(ip => {
      return `<tr><td>A</td><td>${domain}</td><td>${ip}</td></tr>`
    }).join("")

    const nsRows = data.ns_records.map(ns => {
      return `<tr><td>NS</td><td>${domain}</td><td>${ns}</td></tr>`
    }).join("")

    const cnameRows = data.cname_records.map(record => {
      const herokuConfigured = record.cname.includes("heroku");
      const iconClass = herokuConfigured ? "text-success" : "text-danger";
      const iconTitle = herokuConfigured ? "Domaine configuré pour Heroku" : "Domaine non configuré pour Heroku";
      const iconName = herokuConfigured ? "check-circle" : "times-circle";

      return `
        <tr>
          <td>CNAME</td>
          <td>${record.subdomain}.${domain}</td>
          <td>
            ${record.cname}
            <i class="fas fa-${iconName} ${iconClass}" title="${iconTitle}"></i>
          </td>
        </tr>
      `;
    }).join("")

    const mxRows = data.mx_records.map(mx => {
      return `<tr><td>MX</td><td>${domain}</td><td>${mx.priority} ${mx.exchange}</td></tr>`
    }).join("")

    const txtRows = data.txt_records.map(txt => {
      return `<tr><td>TXT</td><td>${domain}</td><td>${txt}</td></tr>`
    }).join("")

    const table = `
      <table class="table table-striped table-bordered">
        <thead>
          <tr>
            <th>Type</th>
            <th>Nom de domaine</th>
            <th>Valeur</th>
          </tr>
        </thead>
        <tbody>
          ${aRows + nsRows + cnameRows + mxRows + txtRows}
        </tbody>
      </table>
    `

    this.tableTarget.innerHTML = table
  }
}

File: app/controllers/dns_check_controller.rb
require 'dnsruby'

class DnsCheckController < ApplicationController
  def check
    domain = params[:domain]

    if domain.present?
      dns_resolver = Dnsruby::Resolver.new

      @a_records = dns_resolver.query(domain, 'A').answer.select { |record| record.type == 'A' }.map(&:address).map(&:to_s)
      @ns_records = dns_resolver.query(domain, 'NS').answer.select { |record| record.type == 'NS' }.map(&:domainname).map(&:to_s)
      @mx_records = dns_resolver.query(domain, 'MX').answer.select { |record| record.type == 'MX' }.map { |record| { priority: record.preference, exchange: record.exchange.to_s } }
      @txt_records = dns_resolver.query(domain, 'TXT').answer.select { |record| record.type == 'TXT' }.map(&:strings).map { |strings| strings.join(" ") }

      # Recherchez les enregistrements CNAME pour les sous-domaines courants
      common_subdomains = ['www', 'mail', 'ftp']
      @cname_records = common_subdomains.map do |subdomain|
        begin
          cname_record = dns_resolver.query("#{subdomain}.#{domain}", 'CNAME').answer.select { |record| record.type == 'CNAME' }.map(&:domainname).map(&:to_s)
          Rails.logger.debug "CNAME record for #{subdomain}.#{domain}: #{cname_record}"
          { subdomain: subdomain, cname: cname_record.first } if cname_record.present?
        rescue Dnsruby::NXDomain
          nil
        end
      end.compact

      render json: {
        a_records: @a_records,
        ns_records: @ns_records,
        cname_records: @cname_records,
        mx_records: @mx_records,
        txt_records: @txt_records
      }
    else
      @dns_records = []
      render :home
    end
  end
end
